hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.signature.hapi;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.Key;
+import com.hedera.hapi.node.base.KeyList;
+import com.hedera.hapi.node.base.ThresholdKey;
+import com.hedera.hapi.node.state.token.Account;
+import com.hedera.node.app.spi.signatures.SignatureVerification;
+import com.swirlds.common.crypto.TransactionSignature;
+import com.swirlds.common.crypto.VerificationStatus;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A {@link Future} that waits on a {@link List} of {@link TransactionSignature}s to complete signature checks, and
+ * yields a {@link SignatureVerification}.
+ */
+class SignatureVerificationFuture implements Future<SignatureVerification> {
```

Comment:
Have we considered using `CompletableFuture` machinery here? It feels like we could get the necessary behavior out-of-the-box.

Reply:
I tried, but didn't find a way in any JDK classes to aggregate multiple futures into one. And then after talking with you I realized that determining "pass/fail" requires walking over the key structure again and comparing which signatures passed/failed to determine whether, for example, a ThresholdKey is passing or not. I'm not sure what to do about it.